packages/evm/src/eof/verify.ts

```diff
@@ -0,0 +1,503 @@
+  terminatingOpcodes.add(0xfe) // INVALID
+
+  terminatingOpcodes.add(0xee) // RETURNCONTRACT
+
+  terminatingOpcodes.add(0xe4) // RETF
+  terminatingOpcodes.add(0xe5) // JUMPF
+
+  terminatingOpcodes.add(0xe0) // RJUMPing back into code section is OK
+
+  for (const opcode of terminatingOpcodes) {
+    if (!opcodeNumbers.has(opcode)) {
+      terminatingOpcodes.delete(opcode)
+    }
+  }
+
+  const validJumps = new Set<number>()
+
+  // Add all reachable code sections
+  const reachableSections: { [key: number]: Set<number> } = {}
+
+  let codeSection = -1
+  for (const code of container.body.codeSections) {
+    codeSection++
+
+    reachableSections[codeSection] = new Set()
+
+    const returningFunction = container.body.typeSections[codeSection].outputs === 0x80
+
+    // Tracking set of reachable opcodes
+    const reachableOpcodes = new Set<number>()
+    reachableOpcodes.add(0)
+
+    // Validate that each opcode is defined
+    let ptr = 0
+    let lastOpcode: number = 0 // Note: code sections cannot be empty, so this number will always be set
+
+    // Implement the EIP 5450 stack validation algorithm
+    const inputs = container.body.typeSections[codeSection].inputs
+    let maxStackHeight = inputs
+    // These arrays track the min/max stack height **before** executing the instruction
+    const stackHeightMin: number[] = [inputs]
+    const stackHeightMax: number[] = [inputs]
+
+    // This loop will loop over the entire code section and will validate various rules
+    // For (most) validation rules, see https://github.com/ipsilon/eof/blob/main/spec/eof.md
+    // For all validation rules per opcode, find the corresponding EIP, the rules are there
+    while (ptr < code.length) {
+      // This set tracks the successor opcodes of this opcode (for stack purposes)
+      const successorSet = new Set<number>()
+
+      // ReachableOpcodes: this can likely be deleted after implementing the 5450 algorithm
+      if (!reachableOpcodes.has(ptr)) {
+        validationError(EOFError.UnreachableCode)
+      }
+
+      if (stackHeightMin[ptr] === undefined || stackHeightMax[ptr] === undefined) {
+        // This error either means that the code is unreachable,
+        // or it is possible that it is only reachable via a backwards jump
+        validationError(EOFError.UnreachableCode)
+      }
+
+      validJumps.add(ptr)
+      const opcode = code[ptr]
+
+      const minStackCurrent = stackHeightMin[ptr]
+      const maxStackCurrent = stackHeightMax[ptr]
+
+      const opcodeInputs = stackDelta[opcode].inputs
+      const opcodeOutputs = stackDelta[opcode].outputs
+
+      if (minStackCurrent - opcodeInputs < 0) {
+        validationError(EOFError.StackUnderflow)
+      }
+
+      const delta = opcodeOutputs - opcodeInputs
+
+      let minStackNext = minStackCurrent + delta
+      let maxStackNext = maxStackCurrent + delta
+
+      if (maxStackNext > 1023) {
+        // TODO verify if 1023 or 1024 is the right constant
```

Comment:
EOF spec seems to says this should be 1023 (assuming this code section is for a complete EOF function)

Reply:
I should likely remove this TODO, I think I changed it to 1024 and then reference tests started to fail.

Reply:
Does this check cover both overflow for `CALLF` and `JUMPF`?

Reply:
This cannot be done at the verification level and has to be done at runtime. This is done (see their definitions in `functions.ts`)